Feature QSPI: add STM32 implementation

0xc0170 opened the pull request:
experimental QSPI stm32 (using disco-f469ni target), sharing here for review to accelerate the progress with this implementation

@ARMmbed/team-st-mcd this is work in progress, there are two PR active that change HAL a bit, so this will be rebased .

I am updating the test example to continue testing this implementaiton, once we have it ready will be shared. What is missing here is polling mode to get status, that one I believe is related to the QSPI custom instruction that is in another PR. To directly talk to the device to read its registers.

If you find a minute to review, and send feedback ?

bcostm commented on targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F469xI/TARGET_DISCO_F469NI/PeripheralPins.c:
```diff
@@ -343,3 +343,24 @@ const PinMap PinMap_CAN_TD[] = {
+};
+
+const PinMap PinMap_QSPI_SSEL[] = {
+    {PB_6,  0, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_SPI2)},
```
It's `GPIO_AF10_QSPI` in ST CubeF4 example

adustm replied:
also in the datasheet ;)

0xc0170 replied:
PF10 is clk, and  PB6 is ssel? I was following the f469 disco example that has the pinout,:

```
/* Definition for QSPI Pins */
#define QSPI_CS_PIN                GPIO_PIN_6
#define QSPI_CS_GPIO_PORT          GPIOB
#define QSPI_CLK_PIN               GPIO_PIN_10
#define QSPI_CLK_GPIO_PORT         GPIOF
#define QSPI_D0_PIN                GPIO_PIN_8
#define QSPI_D1_PIN                GPIO_PIN_9
#define QSPI_D2_PIN                GPIO_PIN_7
#define QSPI_D3_PIN                GPIO_PIN_6
#define QSPI_DX_GPIO_PORT          GPIOF

```

I'll check the datasheet !

adustm commented on targets/TARGET_STM/TARGET_STM32F4/TARGET_STM32F469xI/TARGET_DISCO_F469NI/PeripheralPins.c:
```diff
@@ -343,3 +343,24 @@ const PinMap PinMap_CAN_TD[] = {
+    {PF_7,  0, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF9_QSPI)},
+    {PF_8, 0, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF10_QSPI)},
+    {PF_9,  0, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF10_QSPI)},
+
```
Hello, thank you for this.
The question is: do we split the 4 data lines into 4 separated buffers or not ?
If not, could we use STM_PIN_DATA_EXT to add the Dx value (D0 / D1 / D2 /D3) ?

The bank number information is also required (I suggest to use QSPIBK_1 and QSPIBK_2 the second field) for Data and NSS buffers (not needed for CLK buffer).

Cheers

bcostm replied:
Or use the 2nd field for the peripheral name (QUADSPI) and use other fields for the bank and data line numbers:

```
{PF_6,  QUADSPI, STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF9_QSPI, 1, 3, 0)}, // BK1 / D3
{PF_7,  QUADSPI, STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF9_QSPI, 1, 3, 0)}, // BK1 / D2
```
...
Because on some devices there are 2 banks to control 2 QSPI flash.

0xc0170 replied:
I am missing some context.  can you reference some example or driver that does this - setting the bank + Dx value ?

> Or use the 2nd field for the peripheral name (QUADSPI) and use other fields for the bank and data line numbers:

why peripheral need would be part of the pin configuration?

0xc0170 replied:
It might be easier, if I provide you write access to quickly do the requested changes to my branch so would appear here ?

bcostm replied:
In fact the second field is always used for the peripheral name: ADC_1, PWM_2, SPI_1 etc... My remark was just to do the same for the QSPI.

Example with PWM:
{PE_10, PWM_1, STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM1, 2, 1)},  // TIM1_CH2N
In the STM_PIN_DATA_EXT we use the fourth field to indicate the channel (here 2) and the fifth field used to indicate if the channel is reverted (here value = 1) or not (value = 0).

I was thinking doing the same for the QSPI pins: fourth field used to indicate the bank number and the fifth field to indicate the data line. But thinking again, I am not sure if the data line information is really required? Maybe not after all.

0xc0170 replied:
Not certain what data line hold as information.

Can you send PR to my fork to get this bank info in place, or point me to the directions how this dual bank works and should be defined for pins? Or share the code snippet here?

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+    pinmap_pinout(sclk, PinMap_QSPI_SCLK);
+    pinmap_pinout(ssel, PinMap_QSPI_SSEL);
+
+    if (HAL_QSPI_Init(&obj->handle) != HAL_OK) {
```
Question: what about the DMA configuration as done CubeF4 QSPI example ?

0xc0170 replied:
DMA not at the moment

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+{
+    qspi_status_t status = QSPI_STATUS_OK;
+
+    // TODO calculate prescalers properly, needs some work
```
We can do maybe the same as done in spi_frequency function.

0xc0170 replied:
Thanks for the tip. To use prescaler table and loop until we get the highest available? I can do that

0xc0170 replied:
I cant find any prescaler values for QSPI as it is for SPI.

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+    if (HAL_QSPI_Init(&obj->handle) != HAL_OK) {
+        status = QSPI_STATUS_ERROR;
+    }
+    return QSPI_STATUS_OK;
```
return status;

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+
+    obj->handle.Instance = QUADSPI;
+
+    // TODO pinmap here for pins (enable clock)
```
We will have also to check if the data and NSS pins are on the same bank (BK1 or BK2 when it exists)

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+
+    obj->handle.Init.ClockMode = mode == 0 ? QSPI_CLOCK_MODE_0 : QSPI_CLOCK_MODE_3;
+
+    obj->handle.Instance = QUADSPI;
```
The instance name can be retrieved from the PeripheralPins.c. Not really mandatory as there is only one instance but to be aligned with other peripherals.

0xc0170 replied:
Is there STM32 that has two or more QSPI?

bcostm replied:
For what I have seen so far, there is one instance. But some devices has two banks to communicate with two different flash for example.

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+void qspi_prepare_command(const qspi_command_t *command, QSPI_CommandTypeDef *st_command) 
+{
+    // TODO: shift these around to get more dynamic mapping
+    switch (command->instruction.bus_width) {
```
Add ?
```
default:
st_command->InstructionMode = QSPI_INSTRUCTION_NONE;
```

0xc0170 replied:
fixed

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+            st_command->AddressMode = QSPI_ADDRESS_4_LINES;
+            break;
+        default:
+            st_command->AddressMode = QSPI_ADDRESS_1_LINE;
```
Use `QSPI_ADDRESS_NONE `instead as default ?

0xc0170 replied:
I am proposing to change hal to have NONE for bit_width, not in the size. I'll rework this and would be fixed.

For now, I'll set defalt to none.

0xc0170 replied:
fixed

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+            break;
+    }
+
+    if (command->address.size == QSPI_CFG_ADDR_SIZE_NONE) {
```
Remove this ? if the ADDRESS_NONE is managed above ?

0xc0170 replied:
will do cleanup once I get bit_width NONE setting in place.

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_4_LINES;
+            break;
+        default:
+            st_command->AlternateByteMode = QSPI_ALTERNATE_BYTES_1_LINE;
```
Use `QSPI_ALTERNATE_BYTES_NONE ` instead as default ?

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+            break;
+    }
+
+    if (command->alt.size == QSPI_CFG_ALT_SIZE_NONE) {
```
Remove this ? if the QSPI_ALTERNATE_BYTES_NONE is managed above ?

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+            st_command->DataMode = QSPI_DATA_4_LINES;
+            break;
+        default:
+            st_command->DataMode = QSPI_DATA_1_LINE;
```
Use QSPI_DATA_NONE instead as default ?

bcostm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,245 @@
+
+    if (HAL_QSPI_Command(&obj->handle, &st_command, HAL_QPSI_TIMEOUT_DEFAULT_VALUE) != HAL_OK) {
+        status = QSPI_STATUS_ERROR;
+        return status;
```
`return `to be removed

0xc0170 replied:
the intention is to early leave if command fails.

0xc0170 commented:
/morph build

0xc0170 commented:
/morph build

theotherjimmy commented:
Needs a rebase. @0xc0170 Please remove the needs work label after the rebase.

0xc0170 commented:
Rebased, contains latest updates. I added also QSPI_1 as requested (now data pins are checked if they come within the same peripheral and used later  to get instance).

I am still failing to read register with disco f469ni from the memory device 😞

0xc0170 commented:
I reworked command_tranfer function (it now  contains 3 phases, only command to be sent, tx and rx), please review.

0xc0170 commented:
/morph build

0xc0170 commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,254 @@
+
+    obj->handle.Init.ClockMode = mode == 0 ? QSPI_CLOCK_MODE_0 : QSPI_CLOCK_MODE_3;
+
+    QSPIName qspi_data_io0 = (QSPIName)pinmap_peripheral(io0, PinMap_QSPI_DATA);
```
io1 is right below, line 141? yes the check is missing io1, I'll fix that, bad refactor :-)

0xc0170 commented:
Current status, my small test fails to read from memory bytes that were written (looks like something with freq/or dummy cycles with the memory, investigating).

bcostm commented:
I don't know if it will help, but concerning dummy cycles, I see this command:
```
  /* Update volatile configuration register (with new dummy cycles) */  
  s_command.Instruction = WRITE_VOL_CFG_REG_CMD;
  MODIFY_REG(reg, N25Q128A_VCR_NB_DUMMY, (N25Q128A_DUMMY_CYCLES_READ_QUAD << POSITION_VAL(N25Q128A_VCR_NB_DUMMY)));
```
In the BSP file we use:
https://os.mbed.com/teams/ST/code/BSP_DISCO_F469NI/file/123b894b49dd/Drivers/BSP/STM32469I-Discovery/stm32469i_discovery_qspi.c/

and the dummy cycles value is:
```
#define N25Q128A_DUMMY_CYCLES_READ_QUAD      10
```

0xc0170 commented:
I'll rebase this one more time once #5665 and #5566 gets in. I would like to run CI, and integrate it to continue with working on top of latest changes.

0xc0170 commented:
/morph build

kegilbert commented:
Had a lot of timeouts in that test run, going to rerun it.
/morph test

0xc0170 commented:
/morph test

adustm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,260 @@
+
+    obj->handle.Init.ClockMode = mode == 0 ? QSPI_CLOCK_MODE_0 : QSPI_CLOCK_MODE_3;
+
+    QSPIName qspi_data_first = (SPIName)pinmap_merge(io0, io1);
```
QSPIName instead of SPIName ? (for the 3 lines)

adustm commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -0,0 +1,260 @@
+    if (qspi_data_first != qspi_data_second || qspi_data_second != qspi_data_third ||
+        qspi_data_first != qspi_data_third) {
+        return QSPI_STATUS_INVALID_PARAMETER;
+    }
```
Hi @0xc0170 
pinmap_merge is supposed to check that we are using pins from the same HW module. It takes the HW name as input parameters and not directly pin names.
The correct implementation would look like
```
    QSPIName qspiio0name = (QSPIName)pinmap_peripheral(io0, PinMap_QSPI_DATA);
    QSPIName qspiio1name = (QSPIName)pinmap_peripheral(io1, PinMap_QSPI_DATA);
    QSPIName qspiio2name = (QSPIName)pinmap_peripheral(io2, PinMap_QSPI_DATA);
    QSPIName qspiio3name = (QSPIName)pinmap_peripheral(io3, PinMap_QSPI_DATA);
    QSPIName qspiclkname = (QSPIName)pinmap_peripheral(sclk, PinMap_QSPI_SCLK);
    QSPIName qspisselname = (QSPIName)pinmap_peripheral(ssel, PinMap_QSPI_SSEL);

    QSPIName qspi_data_first = (QSPIName)pinmap_merge(qspiio0name, qspiio1name);
    QSPIName qspi_data_second = (QSPIName)pinmap_merge(qspiio2name, qspiio3name);
    QSPIName qspi_data_third = (QSPIName)pinmap_merge(qspiclkname, qspisselname);

    if (qspi_data_first != qspi_data_second || qspi_data_second != qspi_data_third ||
        qspi_data_first != qspi_data_third) {
        return QSPI_STATUS_INVALID_PARAMETER;
    }

```

0xc0170 commented:
@adustm Thanks for the pointer, I retrieved older correct version of qspi api implementation here ! I almost
 lost it during last rebases and multiple branches sync 😭

0xc0170 commented:
/morph build

0xc0170 commented:
I'll merge this as it is, more fixes will come via separate patches for the feature branch.
